### PR TITLE
net-mail/poppasswd_ceti: bump 1.8.7, EAPI=7 to cancel last rites

### PR DESCRIPTION
**Pull request by mmogilvi**
LATEST: pushed 2019-10-28 with all requested changes (still true as of 2019-11-02)

----
UPDATE: New version of pull request 2019-10-23 (1.8.7)

----
OLDER VERSION of pull request (1.8.5-r3):

Also incorporate a Ubuntu NULL-check patch, document an
alternative way to to drive it (using sudo), and make the
xinetd dependency conditional on the USE flag.

Bug: https://bugs.gentoo.org/696252
Signed-off-by: Matthew Ogilvie <mmogilvi+gnto@zoho.com>

-----

OR: Even simpler is just to change the EAPI=0 line to EAPI=7.  There
does not appear to be anything in the ebuild that cares.

-----

OR:  If there is some other tool that provides the
same functionality as poppassd, I could switch to the other tool.

Basically, I would need something that a web script (or similar)
can provide with three pieces of information (a username, old password,
and new password), and it securely updates the user's system account
to the new password only if the old password matches.

I haven't found any in-tree options that do that.  Some
near-misses include:

   - https://github.com/courtney-rosenthal/web-chpass
     Looks like it would be a reasonable alternative to poppassd_ceti,
     but it is not currently in portage, and it is overkill for my
     purposes: In addition to a low-level tool that replaces
     poppassd_ceti, it also includes web frontend stuff I would
     rather just keep using what I've already integrated into
     the relevant web site...

   - "chpasswd" (part of sys-apps/shadow) = Intended for scripting, but
     does not use or confirm old password.

   - Drive "passwd": "passwd" specifically makes itself hard to
     script: Uses tty pairs, generally restricted to "current user"
     unless root, and root skips the old password check completely
     (would need to check passwords separately).

   - Lots of other web hits look overly complex and/or have obvious
     security problems.

-----

Minor FYI:

   - There was an ancient security problem from 2005, but
     current versions are OK: https://security.gentoo.org/glsa/200501-22
   - I also use a combination of www-apache/mod_authnz_external and
     www-apache/pwauth just to validate passwords (not change them).

### CODE REVIEW

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+EAPI=7
+
+inherit eutils toolchain-funcs pam portability
+
+MY_PN="poppassd"
+MY_P="${MY_PN}-${PV}"
+S=${WORKDIR}/${MY_P}
+
+DESCRIPTION="Password change daemon with PAM support"
+HOMEPAGE="http://echelon.pl/pubs/poppassd.html"
```
Comment (juippis) [inline net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild]: This didnt work for me. Is it just temporarily down, or gone for good? Any alternatives?

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+DEPEND="sys-libs/pam"
+
+RDEPEND="${DEPEND}
+	xinetd? ( sys-apps/xinetd )
+"
+
+src_compile() {
+	$(tc-getCC) -c ${CFLAGS} ${MY_PN}.c || die "Compile failed."
+	$(tc-getCC) ${LDFLAGS} -o poppassd ${MY_PN}.o -lpam $(dlopen_lib) || die "Linking failed."
+}
```
Comment (juippis) [inline net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild]: Could you check whether this needs AR modified as well? See qmail.eclass or qmail-notify for example.

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	dodoc README
+
+	pamd_mimic_system poppassd auth account password
+
+	insinto /etc/xinetd.d
+	newins "${FILESDIR}"/poppassd.xinetd poppassd
+
+	insinto /usr/sbin
+	insopts -o root -g bin -m 500
+	doins poppassd || die "Install failed."
```
Comment (juippis) [inline net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild]: `|| die` not needed here.

File: profiles/package.mask [hunk 4]
```diff
@@ -164,7 +164,6 @@ net-irc/jircii
 net-irc/miau
 net-mail/hotwayd
 net-mail/popick
-net-mail/poppassd_ceti
```
Comment (juippis) [inline profiles/package.mask]: Please make this change in a separate commit, this file is everliving and will probably change a thousand times before this gets merged. Having it in a separate commit makes my life so much easier :)

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild (hunk 1 above)
Comment (mmogilvi) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild]: I think the link is completely dead.

The only reason I'm updating this ebuild is because of a new policy to apparently either update or remove all old EAPI=0 ebuilds, and I don't want it removed.

Most of the top google results are for the Ubuntu / lauchpad version of this package, or for independent implementations of the same conceptual service.  Nothing looks like a project web page for this package specifically...

I've been using previous -rX version(s) (maybe just 1 version?) on a server for nearly 15 years without worrying about where it came from; it was probably picking it up from gentoo's mirrors.

What's the appropriate thing to do for a URL in this circumstance?

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild (hunk 2 above)
Comment (mmogilvi) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild]: AR shouldn't be needed:  Unlike your qmail examples (which are apparently modifying config files associated with the built-in makefiles/scripts?), this is bypassing the Makefile completely and running the two commands needed to build it directly.  It only has one .c file, and does not build any libraries.

This is the way -r2 works, but I could potentially change it patch the variables in the Makefile instead of bypassing the Makefile.  That would probably be more complicated, but also more flexible (if a user wanted to add more source files using epatch_user).  Is it worth it?

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild (hunk 3 above)
Comment (mmogilvi) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild]: Done.

File: profiles/package.mask (hunk 4 above)
Comment (mmogilvi) [reply profiles/package.mask]: Done.

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild (hunk 1 above)
Comment (Philippe23) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild]: Seems to me that the project moved to Github years ago: https://github.com/kravietz/poppassd-ceti  (And it looks like it's had a release or two since then as well.)

It also adopted autotools since then, which may make your `src_compile()` more standardized.

Comment (mmogilvi) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild]: Thanks.  I didn't search for the correct keywords, and didn't find that (underscore instead of dash, didn't try author name, etc).

I'll look into replacing my pull request with the latest release instead...

Comment (juippis) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.5-r3.ebuild]: Yes please :)

We'd also prefer if you became the maintainer for it, so it won't be left hanging anymore.
https://wiki.gentoo.org/wiki/Project:Proxy_Maintainers/User_Guide#Proxied_maintainer_in_metadata.xml
just edit the metadata.xml file in a separate commit or so.

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+LICENSE="GPL-2"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="+xinetd"
+
+DEPEND="sys-libs/pam"
+
+RDEPEND="${DEPEND}
+	xinetd? ( sys-apps/xinetd )
```
Comment (juippis) [inline net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: Sorry to say, but if this doesn't have a function in this ebuild, it doesnt belong as a USE Flag. Use `optfeature` from eutils.eclass instead, or a-like. 
https://wiki.gentoo.org/wiki/Project:Quality_Assurance/Policies#USE-Controlled_Optional_RDEPENDs
There's also insinto /etc/xinetd.d below, are you sure its optional?

Comment (mmogilvi) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: TLDR version: [x]inetd is essentially never the only way to run ANY package, even if for some packages it is "usually" the only way a package is used.  So it is hard to decide how it should be indicated in package dependencies, especially without some kind of global guidelines.

----
In a quick, informal survey / skim of various ebuilds that interact with inetd, I noticed there is a fair amount of inconsistency and strangeness with respect to how it is handled.  For example:

   * Background: Just by its very nature, generally speaking/in most cases, anything that is intended to be launched by one form of inetd could easily be launched by any other form.  In fact, inetd is usually not actually required at all, even if it is "typical".  It could be executed directly or via sudo or ssh instead, and just talk to it with its appropriate protocol via stdin/stdout.  There might be exceptions to this for various odd reasons, but I believe such exceptions are rare.
   * There is a virtual/inetd, but only a few packages depend on it.  More packages directly depend on sys-apps/xinetd instead, but in most cases the specificity is probably unnecessary.
   * Packages that depend on virtual/inetd generally still install /etc/xinetd.d/* config files, which would generally be ignored if the user elects to use some other inetd (or ssh or sudo or ...).
   * Clearly xinetd is the default (preferred) inetd (better configurability, ability to have a directory of independently-managed config files, etc), but it isn't the only option in the tree.
   * Most packages that use the xinetd USE flag do two things with it: add a conditional dependency (sometimes on sys-apps/xinetd, sometimes on virtual/inetd), and conditionally install the xinetd config file for the package.  (A few packages also tweak something else, but I didn't look carefully to see if any of them change compile-time options, vs only change default configuration files.  I suspect it is mostly default files...)
   * dev-vcs/cvs is an interesting case study.  It does not indicate a dependency on any inetd at all, but will install xinetd config files if cvs is compiled in a way that allows it to act as a server via the "server" USE flag.  (It looks like with the ebuild, cvs is always configured to support client mode, although I could have sworn that the lower level configure can actually setup a server-only cvs (but not as default), at least in the distant past.)  But even when server is enabled, it is (was? cvs is almost dead) more common to invoke cvs server mode via ssh, not via an inetd.
   * I believe there is a policy that systemd unit files are always installed, even if the systemd USE flag is off and systemd itself is masked.  Perhaps arguably xinetd config files are conceptually similar enough that maybe they should fall under the same policy?

I wonder if maybe some more guidelines or policies about inetd should be discussed and adopted across gentoo?  Some possibilities to consider include:

   * Only depend on virtual/inetd, unless there is some weird inherit dependency on xinetd specifically?
   * Install xinetd config files if they could work (see cvs), even if some other inetd is currently selected?
   * Avoid depending on any inetd at all unless that really is the only way the package is typically used (not ssh, sudo, or anything else)?
   * There is (or should be) a policy about this already, but if not, make sure that any xinetd config files are "disabled = yes" by default unless the user explicitly enables them.
   * Make an explicit exception to the "USE flag not used for optional dependencies" policy for the "xinetd" USE flag when a package is usually used with xinetd (easiest to configure), but might sometimes be used without it?

----
However, in the absence of any such policy adjustments, the three most likely ways to address your concerns with my poppassd_ceti ebuild might be either:

   * Eliminate any inetd dependency completely, but still install the xinetd config file.  Unconditionally elog a message about how poppassd has to run as root to work, and various ways to run it that way (including xinetd and sudo).
   * Or keep the USE-controlled dependency, but also make installation of the xinetd config file depend on the same flag.  This is the way the flag is used in most ebuilds that test the flag.  Perhaps depend on virtual/inetd instead of sys-apps/xinetd (not sure)?  Probably also elog a message in this case as well.
   * Or we could keep what poppassd_ceti's previous ebuild does, and unconditionally depend on xinetd.  But I've personally never run it that way - I have xinetd installed via dependencies like this, but don't actually start it or use it for anything.  My password change web page backend uses sudo instead.  It would be nice (but not critical) to be able to install poppassd_ceti without xinetd for use cases like mine.

Do you have a preferred strategy?

Comment (mmogilvi) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: In the absence of any guidelines or suggestions, this new PR version goes with my first strategy from near the bottom of the above long analysis.

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+# settled on "poppassd-ceti" (instead of "poppassd_ceti" or no suffix).
+MY_PN="poppassd-ceti"
+MY_P="${MY_PN}-${PV}"
+S=${WORKDIR}/${MY_P}
+
+DESCRIPTION="Password change daemon with PAM support"
+HOMEPAGE="https://github.com/kravietz/poppassd-ceti"
+SRC_URI="https://github.com/kravietz/${MY_PN}/releases/download/v${PV}/${MY_P}.tar.xz"
+
+LICENSE="GPL-2"
```
Comment (mgorny) [inline net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]:
```
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.```

File: net-mail/poppassd_ceti/files/doc-sudoAlternative-1.8.7.patch [hunk 7]
```diff
@@ -0,0 +1,15 @@
+diff --git a/README.md b/README.md
+index 2c37570..5d01a13 100644
+--- a/README.md
++++ b/README.md
+@@ -18,6 +18,10 @@ Poppassd first authenticates the user with its username and password. The authen
+ 
+ If used with web applications, it's recommended that poppassd is only listening on local network interface (`localhost`). The program does not have any access control restrictions and these need to be implemented using operating system's native daemons, such as `inetd` and `tcpd`.
+ 
++Alternatively, a web application might invoke `poppassd` using the equivalent of `sudo /usr/local/sbin/poppassd`, and avoid listening to network interfaces at all. Sudo will need to be appropriately configured. For example, the following `/etc/sudoers` line might be appropriate:
```
Comment (mgorny) [inline net-mail/poppassd_ceti/files/doc-sudoAlternative-1.8.7.patch]: Patching upstream README seems wrong. If you need to add something, use readme.gentoo thingie.

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild (hunk 6 above)
Comment (mgorny) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: IOW, it's GPL-2+

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+src_install() {
+	dodoc README.md
+
+	pamd_mimic_system poppassd auth account password
+
+	insinto /etc/xinetd.d
+	newins "${FILESDIR}"/poppassd.xinetd poppassd
+
+	insinto /usr/sbin
+	insopts -o root -g bin -m 500
```
Comment (mgorny) [inline net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: Mode 500 sounds like an executable, so any reason not to use `exeinto`, `exeopts` and `doexe`?

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	doins poppassd
+}
+
+pkg_postinst() {
+	elog "/usr/sbin/poppassd has to be run as root to work."
```
Comment (mgorny) [inline net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: This also could fit readme.gentoo thingie.

Comment (mmogilvi) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: I've made all these changes, although I have a couple of possible
concerns about the README.gentoo changes:

   1. The README.md patch has already been submitted upstream and
      accepted into its source control, although it hasn't been
      officially released.

      (Generally, the "is xinetd a dependency" question is
      more general than just Gentoo.)

   2. The elog'ed output (and README.md patch) was a replacement for a
      dependency on xinetd (either an "unconditional but
      not-strictly-necessary RDEPEND", or a "USE-controlled RDEPEND").

      I wrote up an extensive analysis about various options,
      inconsistencies, etc in how xinetd dependencies are handled
      (or should be handled) on Oct 25 above.  (Short version:
      xinetd is essentially NEVER the only way to run any mini-service,
      even in cases where it is the most common, so is it a
      dependency or not?)

      Is README.gentoo your recommended way to handle xinetd
      semi-dependencies?

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild (hunk 6 above)
Comment (mmogilvi) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: Changed to GPL-2+.

Very minor FYI:
I originally considered changing it on my own, but hadn't
because the following considerations left it slightly less
obvious if I should, yet:
   - The actual 1.8.7 tarball doesn't mention a license at all
     (other than in the auto* compatibility boilerplate stuff; not the
     tool itself).
   - The corresponding git tag DOES include the GPL "LICENSE" file (somehow
     it seemed to be left out of the tarball despite this).
   - The comment at the top of the C file was only modified to mention
     a license (GPL and the "or later version" terminology) after
     the 1.8.7 release.
   - The previous ebuild (1.8.5-r2) just says GPL-2.
   - The previous tarball (1.8.5) actually has a BSDish license file,
     with an extra paragraph saying that you could instead ignore the
     BSD part and use the GPL instead.

Comment (mgorny) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: Oh, I'm sorry. I've looked only at GitHub, and didn't think it would change that recently. Many people are mistakenly using GPL-2 instead of GPL-2+, so I'm a bit sensitive on it.

I think in this case GPL-2+ is a reasonable option but you are free to follow your own judgment.

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild (hunk 9 above)
Comment (mgorny) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: If it's been accepted upstream, then the patch is fine -- but please include the original commit message (i.e. get it from git-format patch) or at least upstream commit link.

I'd personally still go for readme.gentoo to avoid extra files. Generally, the idea is that if you have a lot of postinst messages and usually they need to be shown to the user only once, readme.gentoo is the solution. Then the user sees them only once, and can read them from README.gentoo file afterwards.

Comment (mmogilvi) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: I just pushed up another rebase:
  - GPL-2+ (The license is somewhat ambiguous in 1.8.7, but I agree that GPL-2+ is reasonable, and it may prevent needing to change it again in a future bump.)
  - Use readme.gentoo-r1.eclass (For my earlier push today I only pushed a README.gentoo file to install directly; I hadn't realized that you were talking about an eclass until you started talking about extra files and "shown ... only once".  I like the eclass a lot better.)
  - Using doexe, etc (I was just copying the old ebuild and didn't realize there was a better alternative).
  - Not bothering with the README.md patch, since I don't really need to duplicate the sudo information that I'm also describing in the readme.gentoo information.  (It won't hurt in a future bump, though.)

File: net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+KEYWORDS="~amd64 ~x86"
+
+DEPEND="sys-libs/pam"
+RDEPEND="${DEPEND}"
+
+FORCE_PRINT_ELOG=1  # possibly remove in the next bump
+DISABLE_AUTOFORMATTING=1
+DOC_CONTENTS="poppassd is installed, but has to be run as root to work.
+
+Most commonly a front end would require sys-apps/xinetd and
```
Comment (mgorny) [inline net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: Could you reformat this to have consistent line width? Right now it seems to be wrapped semi-randomly.

Comment (mmogilvi) [reply net-mail/poppassd_ceti/poppassd_ceti-1.8.7.ebuild]: done

**Comment by mgorny** [comment]
Thanks. Lemme test it and I'll merge if everything is fine.
